capreolus/searcher/special.py

```diff
@@ -0,0 +1,116 @@
+from collections import defaultdict
+from tqdm import tqdm
+
+from capreolus import ConfigOption, Dependency, constants
+from capreolus.utils.loginit import get_logger
+from capreolus.utils.trec import load_trec_topics, topic_to_trectxt
+
+from . import Searcher
+from .anserini import BM25
+
+logger = get_logger(__name__)
+
+
+class MsmarcoPsgSearcherMixin:
+    @staticmethod
+    def convert_to_trec_runs(msmarco_top1k_fn, train):
+        runs = defaultdict(dict)
+        with open(msmarco_top1k_fn, "r", encoding="utf-8") as f:
+            for line in f:
+                if train:
+                    qid, pos_pid, neg_pid = line.strip().split("\t")
+                    runs[qid][pos_pid] = len(runs.get(pos_pid, []))
+                    runs[qid][neg_pid] = len(runs.get(neg_pid, []))
+                else:
+                    qid, pid, _, _ = line.strip().split("\t")
+                    runs[qid][pid] = len(runs.get(qid, []))
+        return runs
+
+    @staticmethod
+    def get_fn_from_url(url):
+        return url.split("/")[-1].replace(".gz", "").replace(".tar", "")
+
+    def download_and_prepare_train_set(self, tmp_dir):
+        url = "https://msmarco.blob.core.windows.net/msmarcoranking/qidpidtriples.train.full.tsv.gz"
+        extract_file_name = self.get_fn_from_url(url)
+        extract_dir = self.benchmark.collection.download_and_extract(url, tmp_dir, expected_fns=extract_file_name)
+        runs = self.convert_to_trec_runs(extract_dir / extract_file_name, train=True)
+        return runs
+
+
+@Searcher.register
+class MsmarcoPsg(Searcher, MsmarcoPsgSearcherMixin):
+    module_name = "msmarcopsg"
+    dependencies = [Dependency(key="benchmark", module="benchmark", name="msmarcopsg")]
+
+    def _query_from_file(self, topicsfn, output_path, cfg):
+        """ only query results in dev and test set are saved """
+        final_runfn = output_path / "searcher"
+        final_donefn = output_path / "done"
+        if os.path.exists(final_donefn):
+            return output_path
+
+        tmp_dir = self.get_cache_path() / "tmp"
+        tmp_dir.mkdir(exist_ok=True, parents=True)
+        output_path.mkdir(exist_ok=True, parents=True)
+
+        # train
+        train_run = self.download_and_prepare_train_set(tmp_dir=tmp_dir)
+        self.write_trec_run(preds=train_run, outfn=final_runfn, mode="wt")
+
+        # dev and test
+        dev_test_urls = [
+            "https://msmarco.blob.core.windows.net/msmarcoranking/top1000.dev.tar.gz",
+            "https://msmarco.blob.core.windows.net/msmarcoranking/top1000.eval.tar.gz",
+        ]
+        runs = {}
+        for url in dev_test_urls:
+            extract_file_name = self.get_fn_from_url(url)
+            extract_dir = self.benchmark.collection.download_and_extract(url, tmp_dir, expected_fns=extract_file_name)
+            runs.update(self.convert_to_trec_runs(extract_dir / extract_file_name, train=False))
+        self.write_trec_run(preds=runs, outfn=final_runfn, mode="a")
```

Comment:
So this searcher does not use the anserini index at all? The "score" of every `(qid, docid)` pair would depend only on the  order in which the doc was added to the run dict (line [27](https://github.com/capreolus-ir/capreolus/compare/master..feature/msmarco_psg#diff-17a2f5012fe850eac26a4193e5d3ccc0f4c76a847064aa331d590e06992cdaadR27) in this file) right? I think I have missed something here :slightly_smiling_face:

Reply:
yeah here i'm just using the provided dev and eval file, mostly for reranking rather than the score of the runfile itself - but u r right, with the bug in line#27 the evaluation on this runfile will not make sense